engine/src/tangl/ir/dispatch.py

```
@@ -0,0 +1,40 @@
+
+@on_materialize(priority=Prio.NORMAL)
+def _structure_item(caller, *, ctx, **_):
+    ...
+
+@on_materialize(priority=Prio.LATE)
+def _post_process_item(caller, *, data, ctx, **_):
+    ...
+
+# story application-level dispatch
+def do_materialize(script_item: BaseScriptItem, *,
+                   ctx: ContextP,
```

Ruff (0.14.10)

Undefined name `ContextP`

(F821)